### ISSUE TITLE
Simple memory placement optimizations not performed when targeting TPUs

### ISSUE THREAD

**Issue by Keno** [tensorflow/tensorflow#21904]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 17.10
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: r1.9
- **Python version**: N/A
- **Bazel version (if compiling from source)**: 0.11.1
- **GCC/Compiler version (if compiling from source)**: 7.2.0
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

### Describe the problem

I'm playing with compiling Julia code to TF graphs for TPU offload. At the moment, I have the two functions:
```
f(X, Y) = Y * (X + c_1*X) * (X + c_2*X) * (X + c_3*X) * (X + c_4*X) * (X + c_5*X) * (X + c_6*X) * (X + c_7*X) * (X + c_8*X)
g(X, Y) = Y * (X + c_1*X) * (X + c_2*X) * (X + c_3*X) * (X + c_4*X) * (X + c_5*X) * (X + c_6*X) * (X + c_7*X) * (X + c_8*X) * (X + c_9 * X)
```
(Note the `*` is matrix-multiply between two matrices). Now, when I run (the TF graph resulting from the compilation of) these two functions against a cloud TPU (Tensorflow version 1.9), with `X`,`Y` both 10,000 x 10,000 Float32 matrices, `c_i` Float32 constants, `f` runs fine, but `g` delivers the following out of memory error:

```
Tensorflow error: Status: Failed to allocate request for 385.74MiB (404480000B) on device ordinal 0

Total hbm usage >= 7.88G:
    reserved                 528.00M
    persistent allocations     7.37G (7.4% fragmentation)
    program                    33.0K

Persistent allocations include some or all of:
    arguments                771.48M (98.9% utilization)
    output                   385.75M (98.9% utilization) [may share some memory with arguments]
```

This seems to indicate to me that it's trying to reserve space for all the intermediate matrix results in high bandwidth memory. I would have expected (and @learyg mentioned that it should) that XLA would have optimized the memory placement in order to avoid this problem. 

Looking at the profile output for the `f` function, it seems to be doing all the broadcasts in order, followed by grouping all the matmuls (which would indeed blow the memory budget):
![screen shot 2018-08-27 at 1 02 37 pm](https://user-images.githubusercontent.com/1291671/44673466-a087b800-a9f9-11e8-9121-86311b5fe40f.png)

I also note that gap between the broadcasts and the matmuls, during which time it seems to be running XLA again. To me that indicates that XLA is not getting the full TF graph on the server side and thus cannot optimize the memory placement properly. I'm hoping that there's just some missing setting in the TF graph def that would allow it to do its job.

### Source code / logs
I was able to reproduce this in raw Tensorflow.jl (which should have a more or less 1:1 API correspondence to python, though of course there may be a bug in the bindings also - cc @malmaud @oxinabox), using code like this:
```
using TensorFlow
using TensorFlow: Device

sess = Session(Graph(); target="grpc://localhost:8470")

with_device(TensorFlow.NamedDevice("/job:tpu_worker/replica:0/task:0/device:TPU:0")) do
    global X,Y,Z
    X = placeholder(Float32, shape=[10000, 10000])
    Z = Y = placeholder(Float32, shape=[10000, 10000])
    for i = 1:8 # Fails for 1:9
        Z *= X + rand(Float32)*X
    end
end

x, z = rand(Float32, 10000, 10000), rand(Float32, 10000, 10000)
run(sess, Z, Dict(X=>x, Y=>z))
```
A serialized GraphDef of the graph for `f` is at https://gist.github.com/Keno/9321e0f1f278b04fbf3b878e551b3ecb

Is there something missing from this GraphDef that would enable XLA to perform the appropriate memory placement optimization on the device?

**Comment by skye** [comment]
@jpienaar are the right person to look at this, or can you redirect to someone who is? Thanks.

**Comment by jpienaar** [comment]
Sure, I'll start looking at it and then redirect as needed.

@Keno  Do you see the the compile op in the trace?

**Comment by Keno** [comment]
Yes, I do. I can probably get you the trace somehow if you tell me what files you need.

**Comment by jpienaar** [comment]
I don't know how the Julia TF integration works, are you currently generating the graphdef directly?

I think the XLA HLO computations, it would be great if you could dump those too (would provide extra validation).

**Comment by Keno** [comment]
Yes, right now we're generating the graphdef directly (in particular the one I attached to the original issue) and then XLA is generated server side. Now that https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/xrt is in tree, I'm hoping to move to that for running stuff on TPUs as soon as that becomes available in the TF version deployed on Cloud TPUs, but for now it's raw TF graphdefs.

**Comment by jpienaar** [comment]
SG, what I would suggest is to generate the graph to look exactly like one would see when using tpu.rewrite. In particular adding the _tpu_replicate attribute to indicate which nodes are part of the same cluster. The easiest would be to rewrite in Python and use that as your template. The graphs look quite different (and the memory usage is much better too).

If you are generating the graphdef directly then I think that would be your best bet for now: it should not require too many changes (adding replicate metadata/input/output nodes and replicate attribute per node in computation cluster) and then you should end up triggering the same extraction passes.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jpienaar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Keno** [comment]
Not an issue when using xrt (which just became available with 1.11), so I consider this resolved.
